src/box/iproto.cc

```diff
@@ -1851,7 +2226,88 @@ net_send_msg(struct cmsg *m)
 {
 	struct iproto_msg *msg = (struct iproto_msg *) m;
 	struct iproto_connection *con = msg->connection;
+	uint32_t stream_id = msg->header.stream_id;
+	mh_int_t pos;
+	struct stream *stream;
+
+	rlist_del_entry(msg, in_pending_list);
+	/*
+	 * If connection was closed, we stop processed remaining
+	 * stream messages, they will be deleted with stream in
+	 * `net_finish_disconnect`
+	 */
+	if (stream_id == 0 || con->state == IPROTO_CONNECTION_CLOSED)
+		goto send_msg;
+
+	pos = mh_streams_find(con->streams, stream_id, 0);
+	assert(pos != mh_end(con->streams));
+	stream = *mh_streams_node(con->streams, pos);
+
+	if (rlist_empty(&stream->pending_list)) {
+		/*
+		 * If no more messages for the current stream
+		 * and no transaction started, then delete it.
+		 */
+		if (stream->txn == NULL) {
+			mh_streams_remove(con->streams, (const struct stream **)
+					  &stream, 0);
+			iproto_stream_delete(stream);
+		}
+	} else {
+		/*
+		 * If there are new messages for this stream
+		 * then schedule their processing.
+		 */
+		struct iproto_msg * next =
+			rlist_first_entry(&stream->pending_list,
+					  struct iproto_msg,
+					  in_pending_list);
+		assert(next != NULL);
+		/*
+		 * It's very important to update wpos before push message
+		 * to tx thread. Otherwise, there may be a situation where
+		 * we reset obuf, but not resetting the corresponding svp.
+		 * For example we have two streams each of which processes
+		 * transaction. And there are several messages in each stream
+		 * which are waited in `pending_list`. For example we have the
+		 * following sequence of requests:
+		 * (0) Some previous requests which was ended long ago.
+		 * (1) Begin in stream_1
+		 * (2) Begin in stream_2
+		 * (3) Some requests in stream_1 and stream_2
+		 * (4) Commit in stream_1
+		 * (5) Commit in stream 2
+		 * (6) Request in stream_1
+		 * All previous requests came at the same time and have the
+		 * same wpos. When (1) starts to be processed in tx thread, he
+		 * reset obuf, which was used in (0). Svp of this obuf was
+		 * previously reseted in `iproto_flush`. All this requests will
+		 * use the same one obuf.
+		 * (7) Request in stream_2
+		 * Request (7) comes later, after one of the previous requests
+		 * is processed, so it can have (and has for our example) other
+		 * wpos. If transaction in stream_1 conflicts with transaction
```

Comment:
I bet this happens not only in case of conflict. It could be select/replace or just calls with short and long functions.